pkg/ipmi/oem.go

```diff
@@ -0,0 +1,333 @@
+	Index             uint8
+	ParameterSelector uint8
+	DIMMPresent       uint8
+	NodeNumber        uint8
+	ChannelNumber     uint8
+	DIMMNumber        uint8
+}
+
+func (DimmDataParam1 *oemDimmLocationDataParam1) data() {}
+
+type oemDimmTypeDataParam2 struct {
+	ManufacturerID    [3]uint8
+	Index             uint8
+	ParameterSelector uint8
+	DIMMType          uint8
+}
+
+func (DimmDataParam2 *oemDimmTypeDataParam2) data() {}
+
+type oemDimmSpeedDatapParam3 struct {
+	ManufacturerID    [3]uint8
+	Index             uint8
+	ParameterSelector uint8
+	DIMMSpeed         uint16
+	DIMMSize          uint32
+}
+
+func (DimmDataParam3 *oemDimmSpeedDatapParam3) data() {}
+
+type oemDimmModulePartNumberDataParam4 struct {
+	ManufacturerID    [3]uint8
+	Index             uint8
+	ParameterSelector uint8
+	ModulePartNumber  [20]byte
+}
+
+func (DimmDataParam4 *oemDimmModulePartNumberDataParam4) data() {}
+
+type oemDimmModuleSerialNumberDataParam5 struct {
+	ManufacturerID     [3]uint8
+	Index              uint8
+	ParameterSelector  uint8
+	ModuleSerialNumber uint32
+}
+
+func (DimmDataParam5 *oemDimmModuleSerialNumberDataParam5) data() {}
+
+type oemDimmModuleManufacturerIDDataParam6 struct {
+	ManufacturerID          [3]uint8
+	Index                   uint8
+	ParameterSelector       uint8
+	ModuleManufacturerIDLSB uint8
+	ModuleManufacturerIDMSB uint8
+}
+
+func (DimmDataParam6 *oemDimmModuleManufacturerIDDataParam6) data() {}
+
+type BoardManufacturerID struct {
+	ID1 uint8
+	ID2 uint8
+	ID3 uint8
+}
+
+var BoardManufacturer = map[string]BoardManufacturerID{
+	"Wiwynn": {ID1: 0x0,
+		ID2: 0x9c,
+		ID3: 0x9c,
+	},
+}
+
+func (i *IPMI) SendOemIpmiCommand(netfn byte, cmd byte, data oemData) error {
+	req := &req{}
+	req.msg.cmd = netfn
+	req.msg.netfn = cmd
+
+	buf := &bytes.Buffer{}
+
+	if err := binary.Write(buf, binary.LittleEndian, data); err != nil {
+		return err
+	}
+
+	req.msg.data = unsafe.Pointer(&buf.Bytes()[0])
+	req.msg.dataLen = uint16(len(buf.Bytes()))
+
+	_, err := i.sendrecv(req)
+
+	return err
+}
+
+// IpmiOemSetProcessorAndDimmInfo sends the information of proccessors and DIMMs to BMC
+func (i *IPMI) IpmiOemSetProcessorAndDimmInfo(si *smbios.Info) error {
+	ProcRequestDataParam1 := oemProcProductNameDataParam1{}
+	ProcRequestDataParam2 := oemProcBasicInfoDataParam2{}
+	DIMMRequestDataParam1 := oemDimmLocationDataParam1{}
+	DIMMRequestDataParam2 := oemDimmTypeDataParam2{}
+	DIMMRequestDataParam3 := oemDimmSpeedDatapParam3{}
+	DIMMRequestDataParam4 := oemDimmModulePartNumberDataParam4{}
+	DIMMRequestDataParam5 := oemDimmModuleSerialNumberDataParam5{}
+	DIMMRequestDataParam6 := oemDimmModuleManufacturerIDDataParam6{}
+
+	t1, err := si.GetSystemInfo()
+	if err != nil {
+		return err
+	}
+
+	//-----------------------------------------------------------------------------------------
+	// Start - Set Processor Information(0x10)
+	t4, err := si.GetProcessorInfo()
+	if err != nil {
+		return err
+	}
+
+	for name, id := range BoardManufacturer {
```

Comment:
Why not `ok, id := BoardManufacturer[name]` instead of iterating over the map?

Reply:
Done.